### PR TITLE
Enable python 3.12

### PR DESCRIPTION
**Pull request by dudoslav**
Enable python 3.12 for this repo

---
[sc-37726]

### CODE REVIEW

**Comment by johnkerl** [comment]
See also #1865

**Comment by kounelisagis** [comment]
Reasons why tests were failing:
- tarfile supports extraction filters, a mechanism to mitigate some of the security issues: https://docs.python.org/3.12/library/tarfile.html#extraction-filters. Solved it using `filter="fully_trusted"`argument for `tf.extractall()`
- `__repr__` of `collections.OrderedDict` for py3.12 has changed: https://github.com/python/cpython/pull/101661. This means we are getting different expected output for 3.12. Solved it using ellipsis marker: https://docs.python.org/3/library/doctest.html#doctest.ELLIPSIS

**Review by kounelisagis — APPROVED** [review]
LGTM, maybe wait for another approval to check my changes.

**Comment by dudoslav** [comment]
Please include pytest outputs for 3.12 executions

File: tiledb/tests/test_compat.py
Comment (ihnorton) [inline tiledb/tests/test_compat.py:50]: Should this order be inverted? ie only use `fully_trusted` if it fails?

Comment (ihnorton) [reply tiledb/tests/test_compat.py:50]: Hmm, actually looking at the docs I don't think we need both - we control these entirely, so just use `fully_trusted` by default (unless there's some way it could fail...)

**Review by ihnorton — APPROVED** [review]
Minor comment about `fully_trusted`; otherwise LGTM. Thanks @kounelisagis @dudoslav!

File: tiledb/tests/test_compat.py
Comment (ihnorton) [reply tiledb/tests/test_compat.py:50]: Oh I see - probably the keyword argument is not valid in those versions. Disregard.

**Comment by kounelisagis** [comment]
> Please include pytest outputs for 3.12 executions

```
Python 3.12.3 | packaged by conda-forge | (main, Apr 15 2024, 18:20:11) [MSC v.1938 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import tiledb
>>> tiledb.version()
(0, 28, 1, 'dev10')
>>> tiledb.libtiledb.version()
(2, 20, 1)
```

File: tiledb/query_condition.py
Comment (teo-tsirpanis) [inline tiledb/query_condition.py:282]: Aren't these redundant?

Comment (kounelisagis) [reply tiledb/query_condition.py:282]: ohh wow

**Pull request merged**
